Python/Algorithms/data_structures/binary_tree.py

```diff
@@ -0,0 +1,214 @@
+
+        '''
+        if node is None:
+            node = self.root
+        
+        fn(node.value)
+        if node.left:
+            self.preorder(node.left,fn)
+        if node.right:
+            self.preorder(node.right,fn)
+    
+    def postorder(self, node = None, fn = print):
+        '''
+        Applies `fn` to all elements during postorder traversal
+
+        '''
+        if node is None:
+            node = self.root
+        
+        if node.left:
+            self.postorder(node.left,fn)
+
+        if node.right:
+            self.postorder(node.right,fn)
+
+        fn(node.value)
+    
+    def levelorder(self, node = None, fn = print):
+        '''
+        Applies `fn` to all elements during level order traversal
+        '''
+        node = node if node else self.root
+        h = self.height(node)
+        for i in range(1,h+1):
+            self.apply_to_level(node, i, print)
+
+    def apply_to_level(self, node, level, fn = print):
+        '''
+        Applies given function to a level.
+        '''
+        if node is None:
+            return
+        if level == 1:
+            fn(node.value)
+        if level > 1:
+            self.apply_to_level(node.left, level - 1, fn)
+            self.apply_to_level(node.right, level - 1, fn)
+
+    
+    def height(self, node):
+        '''
+        find the height of a node in the tree from the bottom most element in that subtree.
+
+        '''
+        if node is None:
+            return 0
+        
+        lheight = self.height(node.left)
+        rheight = self.height(node.right)
+
+        if lheight > rheight:
+            return lheight + 1
+        else:
+            return rheight + 1
+
+    def dfs(self, v):
+        '''
+        searches for element v using dfs in tree
+        '''
+        found = False
+        stack = Stack_List(self.root)
+        while not stack.empty():
+            node = stack.remove()
+            if node.value == v:
+                found = True
+                break
+
+            if node.right:
+                stack.insert(node.right)
+            if node.left:
+                stack.insert(node.left)
+        return found
+    
+    def bfs(self, v):
+        '''
+        searches for element v using bfs in tree
+        '''
+        found = False
+        queue = Queue_List(self.root)
+        while not queue.empty():
+            node = queue.remove()
+
+            if node.value == v:
+                found = True
+                break
+            
+            if node.left:
+                queue.insert(node.left)
+
+            if node.right:
+                queue.insert(node.right)
+        return found
+
+    def topological_sort(self):
+        sortedElements = []
+        self.preorder(self.root, lambda x: sortedElements.append(x))
+        return sortedElements
+
+
+    def __len__(self):
+        return self.length
+    
+
+
+class Tree(CommonTreeMethods,Abstract_Tree):
+    def insert(self,value):
```

Comment:
Level order traversal of a bst is BFS right ? Why don't you use BFS method

Reply:
I never thought of it. I'll implement it using BFS.